### ISSUE TITLE
Use demo_cli with multiple input mp3 files.

### ISSUE THREAD

**Issue by lavarith** [CorentinJ/Real-Time-Voice-Cloning#892]
I have a directory of 50-60 audio samples from the speaker I'm trying to clone. The demo_cli.py script seems to focus on having a single input file to create an embedding. I'm trying to understand what the expected path is when using the CLI to create a voice clone when you have multiple samples. Is there an example where you give a directory full of MP3s?

I can see in the toolbox UI that you can record over and over to better the clone you've created. I assume there's just something about how the encoder or synthesizer works that I'm not understanding. For example, does `embed = encoder.embed_utterance(...)` maintain some ongoing state which I can save after I've embedded the 50 files I have?

Any help would be appreciated.

**Comment by lavarith** [comment]
Here's some of the code I've already written for loading the audio files.

```
## Get input mp3s
audiofiles = [os.path.join(args.sample_path, f) for f in listdir(args.sample_path) if isfile(join(args.sample_path, f))]
print(audiofiles)
    
embeds = []
for audiofile in audiofiles:
    in_fpath = Path(audiofile)

    # Load from filepath
    preprocessed_wav = encoder.preprocess_wav(in_fpath)

   # Then we derive the embedding. There are many functions and parameters that the 
   # speaker encoder interfaces. These are mostly for in-depth research. You will typically
   # only use this function (with its default parameters):
   embed = encoder.embed_utterance(preprocessed_wav)
   print("Created the embedding")
        
   embeds.push(embed)
```

**Comment by blue-fish** [comment]
> I can see in the toolbox UI that you can record over and over to better the clone you've created.

Loading an audio in the toolbox does not train or alter the model in any way. It just changes the speaker embedding (an array of size (256,)) input to the synthesizer. Only the embedding from the last loaded audio is used during generation. Due to dropout, the synthesizer output is never the same for a given set of inputs (speaker embedding + text) **unless** the "random seed" parameter is set to force deterministic generation.

> For example, does `embed = encoder.embed_utterance(...)` maintain some ongoing state which I can save after I've embedded the 50 files I have?

The encoder is deterministic and doesn't learn when multiple samples are loaded. The toolbox output only depends on the last utterance loaded. Try setting "random seed" to a fixed number and you'll find the results will be consistent.

> I have a directory of 50-60 audio samples from the speaker I'm trying to clone.

For better results, you may wish to try the finetuning technique I describe in #437.

**Comment by lavarith** [comment]
Ok. So, my options are then to concatenate a set of mp3 files together to create more audio for the speaker (and set a seed) or to use the finetuning method you linked to.

Curious, why does it say on #437 that it's no longer voice cloning?

**Comment by blue-fish** [comment]
You can also try all of your audio files and pick the one the gives the best results. It's not recommended to pass an extremely long input to the speaker encoder, because it will just average out the features.

> Curious, why does it say on #437 that it's no longer voice cloning?

When using pretrained synthesizer model, the voice you are cloning is unseen. By finetuning the synthesizer, you are allowing it to see the voice you want it to clone, and adjust model weights for better performance. This distinction might not be important to you as an end user.

**Comment by lavarith** [comment]
Thanks for sharing the thought. The nuance is definitely interesting.
